### PR TITLE
Adding Travel Times Historical function

### PR DESCRIPTION
**Pull request by friendchris**
initial commit for new travel times over time function

### CODE REVIEW

File: server/chalicelib/data_funcs.py [hunk 1]
```diff
@@ -108,3 +111,47 @@ def alerts(day, params):
...
+    delta = edate - sdate       # as timedelta
+    
+    # get a range of dates
+    for i in range(delta.days + 1):
+        today = sdate + datetime.timedelta(days=i)
+        today_date = datetime.date(year=today.year, month=today.month, day=today.day)
+        data = travel_times(day=today_date, params=params)
+        for data_dict in data:
+            data_dict['service_date'] = today_date
+        all_data = all_data+data
```
Comment (mathcolo) [inline server/chalicelib/data_funcs.py:138]: `.append` here might be faster?

File: server/chalicelib/data_funcs.py [hunk 2]
```diff
@@ -108,3 +111,47 @@ def alerts(day, params):
                 "text": alert_version["header_text"]
             })
     return flat_alerts
+
+def peak_off_peak(df):
```
Comment (mathcolo) [inline server/chalicelib/data_funcs.py:115]: could this function name be more clear?

File: server/chalicelib/data_funcs.py [hunk 3]
```diff
@@ -108,3 +111,47 @@ def alerts(day, params):
...
+    cal = USFederalHolidayCalendar()
```
Comment (mathcolo) [inline server/chalicelib/data_funcs.py:116]: State holidays like Patriots' Day need to be added to this as well, or does it not matter?

File: server/chalicelib/data_funcs.py [hunk 4]
```diff
@@ -108,3 +111,47 @@ def alerts(day, params):
...
+    holidays = cal.holidays(start=df['dep_dt'].min(), end=df['dep_dt'].max())
+    
+    df['holiday'] = df['dep_dt'].dt.date.astype('datetime64').isin(holidays.date)
+    df['weekday'] = df['dep_dt'].dt.dayofweek
+    
+    conditions  = [(df['holiday']==False) & (df['weekday'] < 5) & ((df['dep_time']>=datetime.time(6,30,0)) & (df['dep_time']<datetime.time(9,0,0))), (df['holiday']==False) & (df['weekday'] < 5) & ((df['dep_time']>=datetime.time(15,30,0)) & (df['dep_time']<datetime.time(18,30,0)))]
```
Comment (mathcolo) [inline server/chalicelib/data_funcs.py:122]: This is a tad verbose but maybe it's ok. Would `np.where` help?

**Review by mathcolo — COMMENTED** [review]
Nice stuff—wondering if it will be slow, but worth trying! Don't forget to fix the linting errors

File: server/chalicelib/data_funcs.py (hunk 1 above)
Comment (friendchris) [reply server/chalicelib/data_funcs.py:138]: .extend!

File: server/chalicelib/data_funcs.py (hunk 2 above)
Comment (friendchris) [reply server/chalicelib/data_funcs.py:115]: train_peak_status?

File: server/chalicelib/data_funcs.py (hunk 3 above)
Comment (friendchris) [reply server/chalicelib/data_funcs.py:116]: no b/c usually the T runs enough service that it's like a normal day

File: server/chalicelib/data_funcs.py (hunk 4 above)
Comment (friendchris) [reply server/chalicelib/data_funcs.py]: np.select is better when you've got >2 outputs (in our case, am peak, pm peak, and off-peak). the conditions are verbose but there are a lot of them! I've broken up the line so it's more clear what is what.

File: server/Pipfile [hunk 5]
```diff
@@ -15,6 +15,8 @@ chalice = "*"
 pytz = "*"
 boto3 = "*"
 flake8 = "*"
+pandas = "*"
+numpy = "*"
```
Comment (mathcolo) [inline server/Pipfile:19]: This is correct! Don't worry about the lock file for now; just make sure to also add this to `requirements.txt`.

Pipfile/Pipfile.lock -> only for local development in Pipenv
requirements.txt -> what Chalice pulls from when packaging everything up for AWS

File: server/chalicelib/data_funcs.py [hunk 6]
```diff
@@ -108,3 +111,52 @@ def alerts(day, params):
...
+    df['weekday'] = df['dep_dt'].dt.dayofweek
+
+    conditions = [(df['holiday'] == 0) & (df['weekday'] < 5) & ((df['dep_time'] >= datetime.time(6, 30, 0)) & (df['dep_time'] < datetime.time(9, 0, 0))),
+                  (df['holiday'] == 0) & (df['weekday'] < 5) & ((df['dep_time'] >= datetime.time(15, 30, 0)) & (df['dep_time'] < datetime.time(18, 30, 0)))]
+    choices = ['am_peak', 'pm_peak']
+    df['peak'] = np.select(conditions, choices, default='off_peak')
+    return df
+
+
+def travel_times_over_time(sdate, edate, params):
```
Comment (mathcolo) [inline server/chalicelib/data_funcs.py:130]: I hate to be that person, but should functions that aggregate data go in a different file?

Comment (friendchris) [reply server/chalicelib/data_funcs.py]: this is complete

**Review by mathcolo — APPROVED** [review]
Think I'm good with this now

**Pull request merged**